src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -409,13 +440,99 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
+                    // that moves the analyzer to a lower priority bucket. However, to ensure that de-prioritization happens for very rare cases,
+                    // we only perform this optimizations when following conditions are met:
+                    //  1. We are performing semantic span-based analysis.
+                    //  2. We are processing 'CodeActionRequestPriority.Normal' priority request.
+                    //  3. Analyzer registers certain actions that are known to lead to high performance impact due to its broad analysis scope,
+                    //     such as SymbolStart/End actions and SemanticModel actions.
+                    //  4. Analyzer did not report a diagnostic on the same line in prior document snapshot.
+
+                    // Conditions 1. and 2.
+                    if (kind != AnalysisKind.Semantic ||
+                        !span.HasValue ||
+                        _priorityProvider.Priority != CodeActionRequestPriority.Normal)
+                    {
+                        return false;
+                    }
+
+                    Debug.Assert(span.Value.Length < _text.Length);
+
+                    // Condition 3.
+                    // Check if this is a candidate analyzer that can be de-prioritized into a lower priority bucket based on registered actions.
+                    if (!await IsCandidateForDeprioritizationBasedOnRegisteredActionsAsync(analyzer).ConfigureAwait(false))
+                    {
+                        return false;
+                    }
+
+                    // Condition 4.
+                    // We do not want to de-prioritize this analyzer if it reported a diagnostic on a prior document snapshot,
+                    // such that diagnostic's start/end lines intersect the current analysis span's start/end lines.
+                    // If an analyzer reported such a diagnostic, it is highly likely that the user intends to invoke the code fix
+                    // for this diagnostic. Additionally, it is also highly likely that this analyzer will report a diagnostic
+                    // on the current snapshot. So, we deem this as an important analyzer that should not be de-prioritized here.
+                    // Note that we only perform this analysis if the prior document, whose existingData is cached, had same number
+                    // of source lines as the current document snapshot. Otherwise, the start/end lines comparison across
+                    // snapshots is not meaningful.
+                    if (existingData.LineCount == _text.Lines.Count &&
+                        !existingData.Items.IsEmpty)
+                    {
+                        _text.GetLinesAndOffsets(span.Value, out var startLineNumber, out var _, out var endLineNumber, out var _);
+
+                        foreach (var diagnostic in existingData.Items)
```

Comment:
is existingData sorted?  binary search might be better here.

Reply:
Code inside the loop is pretty trivial and I don't expect enough diagnostics here to cause performance impact to further optimize this. We can always come back to this later if it shows up in traces.